src/Castle.Windsor/MicroKernel/Arguments.cs

```diff
@@ -17,210 +17,239 @@ namespace Castle.MicroKernel
 
 		public IDictionaryEnumerator GetEnumerator()
 		{
-			return arguments.GetEnumerator();
+			return dictionary.GetEnumerator();
 		}
 
 		public void Remove(object key)
 		{
-			arguments.Remove(key);
+			EnsureWritable();
+			dictionary.Remove(key);
 		}
 
 		public Arguments Clone()
 		{
-			var dictionary = arguments as Dictionary<object, object>;
-			if (dictionary != null)
+			if (dictionary is Dictionary<object, object>)
 			{
-				var comparerExtended = dictionary.Comparer as ArgumentsComparerExtended;
-				if (comparerExtended != null)
-				{
-					return new Arguments(arguments, comparerExtended.CustomComparers);
-				}
+				return new Arguments(dictionary);
 			}
 
-			return new Arguments(arguments);
+			return new Arguments(dictionary);
 		}
 
 		void ICollection.CopyTo(Array array, int index)
 		{
-			arguments.CopyTo(array, index);
+			dictionary.CopyTo(array, index);
 		}
 
 		IEnumerator IEnumerable.GetEnumerator()
 		{
 			return GetEnumerator();
 		}
 
-		private class ArgumentsComparer : IEqualityComparer<object>
+		/// <summary>
+		/// Inserts a new named argument with given key. If an argument for this name already exists, it will be overwritten.
+		/// </summary>
+		public Arguments Insert(string key, object value)
+		{
+			this[key] = value;
+			return this;
+		}
+
+		/// <summary>
+		/// Inserts a new typed argument with given type. If an argument for this type already exists, it will be overwritten.
+		/// </summary>
+		public Arguments Insert(Type dependencyType, object value)
 		{
-			public new virtual bool Equals(object x, object y)
+			this[dependencyType] = value;
+			return this;
+		}
+
+		/// <summary>
+		/// Method that can be used to named/type arguments into <see cref="Arguments"/> from a dictionary.
+		/// </summary>
+		/// <param name="values">Key values are inserted into <see cref="Arguments"/> where the values are used as dependencies</param>
+		/// <returns><see cref="Arguments"/></returns>
+		public Arguments Insert(IDictionary values)
+		{
+			foreach (DictionaryEntry entry in values)
 			{
-				var a = x as string;
-				if (a != null)
+				if (entry.Key is string || entry.Key is Type)
 				{
-					return StringComparer.OrdinalIgnoreCase.Equals(a, y as string);
+					dictionary[entry.Key] = entry.Value;
+				}
+				else
+				{
+					throw new ArgumentException($"The argument key '{entry.Key}' should be of type System.String or System.Type");
 				}
-				return Object.Equals(x, y);
 			}
+			return this;
+		}
 
-			public virtual int GetHashCode(object obj)
+		/// <summary>
+		/// Method that can be used to named arguments into <see cref="Arguments"/> from a read-only dictionary.
+		/// </summary>
+		/// <param name="values"><see cref="IReadOnlyDictionary{S, O}"/> where the dictionary is shallow copied up front into <see cref="Arguments"/> and then used as dependencies</param>
+		/// <returns><see cref="Arguments"/></returns>
+		public Arguments InsertNamed(IReadOnlyDictionary<string, object> values)
```

Comment:
I also think we should mark the methods as virtual, so folks out there can inherit and override our API without having to raise PR's for further type support.